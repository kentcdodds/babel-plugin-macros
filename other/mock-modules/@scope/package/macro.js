// this is used to make sure that you can require macro from node_modules
const {createMacro} = require('../../../src')

const innerFn = jest.fn()
module.exports = createMacro(innerFn)
module.exports.innerFn = innerFn
