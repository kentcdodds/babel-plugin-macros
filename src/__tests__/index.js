import path from 'path'
import {lilconfigSync as lilconfigSyncMock} from 'lilconfig'
import cpy from 'cpy'
import babel from '@babel/core'
import pluginTester from 'babel-plugin-tester'
import plugin from '../'

const projectRoot = path.join(__dirname, '../../')

jest.mock('lilconfig', () => {
  const lilconfigExports = jest.requireActual('lilconfig')
  const actualLilconfigSync = lilconfigExports.lilconfigSync
  function fakeLilconfigSync(...args) {
    fakeLilconfigSync.explorer = actualLilconfigSync(...args)
    return fakeLilconfigSync.explorer
  }
  return {...lilconfigExports, lilconfigSync: fakeLilconfigSync}
})

beforeAll(() => {
  // copy our mock modules to the node_modules directory
  // so we can test how things work when importing a macro
  // from the node_modules directory.
  return cpy(['**/*.js'], path.join('..', '..', 'node_modules'), {
    parents: true,
    cwd: path.join(projectRoot, 'other', 'mock-modules'),
  })
})

beforeEach(() => {
  jest.spyOn(console, 'error').mockImplementation(() => {})
})

afterEach(() => {
  console.error.mockRestore()
  jest.clearAllMocks()
})

expect.addSnapshotSerializer({
  print(val) {
    return (
      val
        .split(projectRoot)
        .join('<PROJECT_ROOT>/')
        .replace(/\\/g, '/')
        // Remove the path of file which thrown an error
        .replace(/Error:[^:]*:/, 'Error:')
    )
  },
  test(val) {
    return typeof val === 'string'
  },
})

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    filename: __filename,
    parserOpts: {
      plugins: ['jsx'],
    },
    generatorOpts: {quotes: 'double'},
  },
  tests: [
    {
      title: 'does nothing to code that does not import macro',
      snapshot: false,
      code: `
        import foo from './some-file-without-macro'

        const bar = require('./some-other-file-without-macro')
      `,
    },
    {
      title: 'does nothing but remove macros if it is unused',
      snapshot: true,
      code: `
        import foo from "./fixtures/eval.macro";

        const bar = 42;
      `,
    },
    {
      title: 'raises an error if macro does not exist',
      error: true,
      code: `
        import foo from './some-macros-that-doesnt-even-need-to-exist.macro'
        export default 'something else'
      `,
    },
    {
      title: 'works with import',
      code: `
        import myEval from './fixtures/eval.macro'
        const x = myEval\`34 + 45\`
      `,
    },
    {
      title: 'works with require',
      code: `
        const evaler = require('./fixtures/eval.macro')
        const x = evaler\`34 + 45\`
      `,
    },
    {
      title: 'works with require destructuring',
      code: `
        const {css, styled} = require('./fixtures/emotion.macro')
        const red = css\`
          background-color: red;
        \`

        const Div = styled.div\`
          composes: \${red}
          color: blue;
        \`
      `,
    },
    {
      title: 'works with require destructuring and aliasing',
      code: `
        const {css: CSS, styled: STYLED} = require('./fixtures/emotion.macro')
        const red = CSS\`
          background-color: red;
        \`

        const Div = STYLED.div\`
          composes: \${red}
          color: blue;
        \`
      `,
    },
    {
      title: 'works with function calls',
      code: `
        import myEval from './fixtures/eval.macro'
        const x = myEval('34 + 45')
      `,
    },
    {
      title: 'Works as a JSXElement',
      code: `
        import MyEval from './fixtures/eval.macro'
        const x = <MyEval>34 + 45</MyEval>
      `,
    },
    {
      title: 'Supports named imports',
      code: `
        import {css as CSS, styled as STYLED} from './fixtures/emotion.macro'
        const red = CSS\`
          background-color: red;
        \`

        const Div = STYLED.div\`
          composes: \${red}
          color: blue;
        \`
      `,
    },
    {
      title: 'supports compiled macros (`__esModule` + `export default`)',
      code: `
        import {css, styled} from './fixtures/emotion-esm.macro'
        const red = css\`
          background-color: red;
        \`

        const Div = styled.div\`
          composes: \${red}
          color: blue;
        \`
      `,
    },
    {
      title: 'supports macros from node_modules',
      code: `
        import fakeMacro from 'babel-plugin-macros-test-fake/macro'
        fakeMacro('hi')
      `,
      teardown() {
        try {
          // kinda abusing the babel-plugin-tester API here
          // to make an extra assertion
          // eslint-disable-next-line
          const fakeMacro = require('babel-plugin-macros-test-fake/macro')
          expect(fakeMacro.innerFn).toHaveBeenCalledTimes(1)
          expect(fakeMacro.innerFn).toHaveBeenCalledWith({
            references: expect.any(Object),
            source: expect.stringContaining(
              'babel-plugin-macros-test-fake/macro',
            ),
            state: expect.any(Object),
            babel: expect.any(Object),
            isBabelMacrosCall: true,
          })
          expect(fakeMacro.innerFn.mock.calls[0].babel).toBe(babel)
        } catch (e) {
          console.error(e)
          throw e
        }
      },
    },
    {
      title: 'optionally keep imports (variable assignment)',
      code: `
        const macro = require('./fixtures/keep-imports.macro')
        const red = macro('noop');
      `,
    },
    {
      title: 'optionally keep imports (import declaration)',
      code: `
        import macro from './fixtures/keep-imports.macro'
        const red = macro('noop');
      `,
    },
    {
      title:
        'optionally keep imports in combination with babel-preset-env (#80)',
      code: `
        import macro from './fixtures/keep-imports.macro'
        const red = macro('noop')
      `,
      babelOptions: {
        plugins: [
          require.resolve('babel-plugin-transform-es2015-modules-commonjs'),
        ],
      },
    },
    {
      title: 'throws an error if the macro is not properly wrapped',
      error: true,
      code: `
        import unwrapped from './fixtures/non-wrapped.macro'
        unwrapped('hey')
      `,
    },
    {
      title: 'forwards MacroErrors thrown by the macro',
      error: true,
      code: `
        import errorThrower from './fixtures/macro-error-thrower.macro'
        errorThrower('hey')
      `,
    },
    {
      title: 'prepends the relative path for errors thrown by the macro',
      error: true,
      code: `
        import errorThrower from './fixtures/error-thrower.macro'
        errorThrower('hey')
      `,
    },
    {
      title: 'appends the npm URL for errors thrown by node modules',
      error: true,
      code: `
        import errorThrower from 'babel-plugin-macros-test-error-thrower.macro'
        errorThrower('hi')
      `,
    },
    {
      title:
        'appends the npm URL for errors thrown by node modules with a slash',
      error: true,
      code: `
        import errorThrower from 'babel-plugin-macros-test-error-thrower/macro'
        errorThrower('hi')
      `,
    },
    {
      title: 'macros can set their configName and get their config',
      fixture: path.join(__dirname, 'fixtures/config/code.js'),
      teardown() {
        try {
          const babelMacrosConfig = require('./fixtures/config/babel-plugin-macros.config')
          const configurableMacro = require('./fixtures/config/configurable.macro')
          expect(configurableMacro.realMacro).toHaveBeenCalledTimes(1)
          expect(configurableMacro.realMacro.mock.calls[0][0].config).toEqual(
            babelMacrosConfig[configurableMacro.configName],
          )

          configurableMacro.realMacro.mockClear()
        } catch (e) {
          console.error(e)
          throw e
        }
      },
    },
    {
      title:
        'when there is an error reading the config, a helpful message is logged',
      error: true,
      fixture: path.join(__dirname, 'fixtures/config/code.js'),
      setup() {
        jest
          .spyOn(lilconfigSyncMock.explorer, 'search')
          .mockImplementationOnce(() => {
            throw new Error('this is a lilconfig error')
          })
        jest.spyOn(console, 'error').mockImplementationOnce(() => {})
        return function teardown() {
          try {
            expect(console.error).toHaveBeenCalledTimes(1)
            expect(console.error.mock.calls[0]).toMatchSnapshot()
            console.error.mockClear()
          } catch (e) {
            console.error(e)
            console.error.mockClear()
            throw e
          }
        }
      },
    },
    {
      title: 'when there is no config to load, then no config is passed',
      fixture: path.join(__dirname, 'fixtures/config/code.js'),
      setup() {
        jest
          .spyOn(lilconfigSyncMock.explorer, 'search')
          .mockImplementationOnce(() => {
            return null
          })
        return function teardown() {
          try {
            const configurableMacro = require('./fixtures/config/configurable.macro')
            expect(configurableMacro.realMacro).toHaveBeenCalledTimes(1)
            expect(configurableMacro.realMacro.mock.calls[0][0].config).toEqual(
              {},
            )
            configurableMacro.realMacro.mockClear()
          } catch (e) {
            console.error(e)
            throw e
          }
        }
      },
    },
    {
      title: 'when configuration is specified in plugin options',
      pluginOptions: {
        configurableMacro: {
          someConfig: false,
          somePluginConfig: true,
        },
      },
      fixture: path.join(__dirname, 'fixtures/config/code.js'),
      teardown() {
        try {
          const configurableMacro = require('./fixtures/config/configurable.macro')
          expect(configurableMacro.realMacro).toHaveBeenCalledTimes(1)
          expect(configurableMacro.realMacro.mock.calls[0][0].config).toEqual({
            fileConfig: true,
            someConfig: true,
            somePluginConfig: true,
          })
          configurableMacro.realMacro.mockClear()
        } catch (e) {
          console.error(e)
          throw e
        }
      },
    },
    {
      title: 'when configuration is specified in plugin options',
      pluginOptions: {
        configurableMacro: {
          someConfig: false,
          somePluginConfig: true,
        },
      },
      fixture: path.join(__dirname, 'fixtures/config/cjs-code.js'),
      teardown() {
        try {
          const configurableMacro = require('./fixtures/config/configurable.macro')
          expect(configurableMacro.realMacro).toHaveBeenCalledTimes(1)
          expect(configurableMacro.realMacro.mock.calls[0][0].config).toEqual({
            fileConfig: true,
            someConfig: true,
            somePluginConfig: true,
          })
          configurableMacro.realMacro.mockClear()
        } catch (e) {
          console.error(e)
          throw e
        }
      },
    },
    {
      title: 'when configuration is specified incorrectly in plugin options',
      fixture: path.join(__dirname, 'fixtures/config/code.js'),
      pluginOptions: {
        configurableMacro: 2,
      },
      teardown() {
        try {
          const configurableMacro = require('./fixtures/config/configurable.macro')
          expect(configurableMacro.realMacro).toHaveBeenCalledTimes(1)
          expect(configurableMacro.realMacro).not.toHaveBeenCalledWith(
            expect.objectContaining({
              config: expect.any,
            }),
          )
          configurableMacro.realMacro.mockClear()
        } catch (e) {
          console.error(e)
          throw e
        }
      },
    },
    {
      title: 'when a custom isMacrosName option is used on a import',
      pluginOptions: {
        isMacrosName(v) {
          return v.endsWith('-macro.js')
        },
      },
      code: `
        import myEval from './fixtures/eval-macro.js'
        const x = myEval\`34 + 45\`
      `,
    },
    {
      title: 'when a custom isMacrosName option is used on a require',
      pluginOptions: {
        isMacrosName(v) {
          return v.endsWith('-macro.js')
        },
      },
      code: `
        const evaler = require('./fixtures/eval-macro.js')
        const x = evaler\`34 + 45\`
      `,
    },
    {
      title:
        'when plugin options configuration cannot be merged with file configuration',
      error: true,
      fixture: path.join(__dirname, 'fixtures/primitive-config/code.js'),
      pluginOptions: {
        configurableMacro: {},
      },
    },
    {
      title:
        'when a plugin that replaces paths is used, macros still work properly',
      fixture: path.join(
        __dirname,
        'fixtures/path-replace-issue/variable-assignment.js',
      ),
      babelOptions: {
        babelrc: true,
      },
    },
    {
      title: 'Macros are applied in the order respecting plugins order',
      code: `
        import Wrap from "./fixtures/jsx-id-prefix.macro";

        const bar = Wrap(<div id="d1"><p id="p1"></p></div>);
      `,
      babelOptions: {
        presets: [{plugins: [require('./fixtures/jsx-id-prefix.plugin')]}],
      },
    },
  ],
})

/* eslint no-console:0 */
