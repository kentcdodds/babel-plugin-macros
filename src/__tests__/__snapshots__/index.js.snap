// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown plugin Macros are applied in the order respecting plugins order: Macros are applied in the order respecting plugins order 1`] = `

import Wrap from "./fixtures/jsx-id-prefix.macro";

const bar = Wrap(<div id="d1"><p id="p1"></p></div>);

      ↓ ↓ ↓ ↓ ↓ ↓

const bar = Wrap(
  <div id="plugin-macro-d1">
    <p id="plugin-macro-p1"></p>
  </div>,
)


`;

exports[`unknown plugin Supports named imports: Supports named imports 1`] = `

import {css as CSS, styled as STYLED} from './fixtures/emotion.macro'
const red = CSS\`
  background-color: red;
\`

const Div = STYLED.div\`
  composes: \${red}
  color: blue;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const red = 'background-color: red;'
const Div = STYLED.div\`composes: background-color: red;
  color: blue;\`


`;

exports[`unknown plugin Works as a JSXElement: Works as a JSXElement 1`] = `

import MyEval from './fixtures/eval.macro'
const x = <MyEval>34 + 45</MyEval>

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79


`;

exports[`unknown plugin appends the npm URL for errors thrown by node modules with a slash: appends the npm URL for errors thrown by node modules with a slash 1`] = `

import errorThrower from 'babel-plugin-macros-test-error-thrower/macro'
errorThrower('hi')

      ↓ ↓ ↓ ↓ ↓ ↓

Error: babel-plugin-macros-test-error-thrower/macro: not helpful Learn more: https://www.npmjs.com/package/babel-plugin-macros-test-error-thrower

`;

exports[`unknown plugin appends the npm URL for errors thrown by node modules: appends the npm URL for errors thrown by node modules 1`] = `

import errorThrower from 'babel-plugin-macros-test-error-thrower.macro'
errorThrower('hi')

      ↓ ↓ ↓ ↓ ↓ ↓

Error: babel-plugin-macros-test-error-thrower.macro: not helpful Learn more: https://www.npmjs.com/package/babel-plugin-macros-test-error-thrower.macro

`;

exports[`unknown plugin does nothing but remove macros if it is unused: does nothing but remove macros if it is unused 1`] = `

import foo from "./fixtures/eval.macro";

const bar = 42;

      ↓ ↓ ↓ ↓ ↓ ↓

const bar = 42


`;

exports[`unknown plugin forwards MacroErrors thrown by the macro: forwards MacroErrors thrown by the macro 1`] = `

import errorThrower from './fixtures/macro-error-thrower.macro'
errorThrower('hey')

      ↓ ↓ ↓ ↓ ↓ ↓

MacroError: very helpful

`;

exports[`unknown plugin macros can set their configName and get their config: macros can set their configName and get their config 1`] = `

import configured from './configurable.macro'

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`


`;

exports[`unknown plugin optionally keep imports (import declaration): optionally keep imports (import declaration) 1`] = `

import macro from './fixtures/keep-imports.macro'
const red = macro('noop');

      ↓ ↓ ↓ ↓ ↓ ↓

import macro from './fixtures/keep-imports.macro'
const red = macro('noop')


`;

exports[`unknown plugin optionally keep imports (variable assignment): optionally keep imports (variable assignment) 1`] = `

const macro = require('./fixtures/keep-imports.macro')
const red = macro('noop');

      ↓ ↓ ↓ ↓ ↓ ↓

const macro = require('./fixtures/keep-imports.macro')
const red = macro('noop')


`;

exports[`unknown plugin optionally keep imports in combination with babel-preset-env (#80): optionally keep imports in combination with babel-preset-env (#80) 1`] = `

import macro from './fixtures/keep-imports.macro'
const red = macro('noop')

      ↓ ↓ ↓ ↓ ↓ ↓

'use strict'

var _keepImports = require('./fixtures/keep-imports.macro')
var _keepImports2 = _interopRequireDefault(_keepImports)
function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj}
}
const red = (0, _keepImports2.default)('noop')


`;

exports[`unknown plugin prepends the relative path for errors thrown by the macro: prepends the relative path for errors thrown by the macro 1`] = `

import errorThrower from './fixtures/error-thrower.macro'
errorThrower('hey')

      ↓ ↓ ↓ ↓ ↓ ↓

Error: ./fixtures/error-thrower.macro: very unhelpful

`;

exports[`unknown plugin raises an error if macro does not exist: raises an error if macro does not exist 1`] = `

import foo from './some-macros-that-doesnt-even-need-to-exist.macro'
export default 'something else'

      ↓ ↓ ↓ ↓ ↓ ↓

Error: Cannot find module './some-macros-that-doesnt-even-need-to-exist.macro' from '<PROJECT_ROOT>/src/__tests__'

`;

exports[`unknown plugin supports compiled macros (\`__esModule\` + \`export default\`): supports compiled macros (\`__esModule\` + \`export default\`) 1`] = `

import {css, styled} from './fixtures/emotion-esm.macro'
const red = css\`
  background-color: red;
\`

const Div = styled.div\`
  composes: \${red}
  color: blue;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const red = css\`
  background-color: red;
\`
const Div = styled.div\`
  composes: \${red}
  color: blue;
\`


`;

exports[`unknown plugin supports macros from node_modules with scope: supports macros from node_modules with scope 1`] = `

import fakeMacro from '@scope/package/macro'
fakeMacro('hi')

      ↓ ↓ ↓ ↓ ↓ ↓

fakeMacro('hi')


`;

exports[`unknown plugin supports macros from node_modules: supports macros from node_modules 1`] = `

import fakeMacro from 'babel-plugin-macros-test-fake/macro'
fakeMacro('hi')

      ↓ ↓ ↓ ↓ ↓ ↓

fakeMacro('hi')


`;

exports[`unknown plugin throws an error if the macro is not properly wrapped: throws an error if the macro is not properly wrapped 1`] = `

import unwrapped from './fixtures/non-wrapped.macro'
unwrapped('hey')

      ↓ ↓ ↓ ↓ ↓ ↓

Error: The macro imported from "./fixtures/non-wrapped.macro" must be wrapped in "createMacro" which you can get from "babel-plugin-macros". Please refer to the documentation to see how to do this properly: https://github.com/kentcdodds/babel-plugin-macros/blob/master/other/docs/author.md#writing-a-macro

`;

exports[`unknown plugin when a custom isMacrosName option is used on a import: when a custom isMacrosName option is used on a import 1`] = `

import myEval from './fixtures/eval-macro.js'
const x = myEval\`34 + 45\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79


`;

exports[`unknown plugin when a custom isMacrosName option is used on a require: when a custom isMacrosName option is used on a require 1`] = `

const evaler = require('./fixtures/eval-macro.js')
const x = evaler\`34 + 45\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79


`;

exports[`unknown plugin when a plugin that replaces paths is used, macros still work properly: when a plugin that replaces paths is used, macros still work properly 1`] = `

import myEval from '../eval.macro'

const result = myEval\`+('4' + '2')\`

global.result = result

      ↓ ↓ ↓ ↓ ↓ ↓

const result = ('foobar', 42)
global.result = result


`;

exports[`unknown plugin when configuration is specified in plugin options: when configuration is specified in plugin options 1`] = `

import configured from './configurable.macro'

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`


`;

exports[`unknown plugin when configuration is specified in plugin options: when configuration is specified in plugin options 2`] = `

const configured = require('./configurable.macro')

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`


`;

exports[`unknown plugin when configuration is specified incorrectly in plugin options: when configuration is specified incorrectly in plugin options 1`] = `

import configured from './configurable.macro'

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`


`;

exports[`unknown plugin when plugin options configuration cannot be merged with file configuration: when plugin options configuration cannot be merged with file configuration 1`] = `

import configured from './configurable.macro'

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/fixtures/primitive-config/babel-plugin-macros.config.js specified a configurableMacro config of type object, but the the macros plugin's options.configurableMacro did contain an object. Both configs must contain objects for their options to be mergeable.

`;

exports[`unknown plugin when there is an error reading the config, a helpful message is logged 1`] = `
[
  There was an error trying to load the config "configurableMacro" for the macro imported from "./configurable.macro. Please see the error thrown for more information.,
]
`;

exports[`unknown plugin when there is an error reading the config, a helpful message is logged: when there is an error reading the config, a helpful message is logged 1`] = `

import configured from './configurable.macro'

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

Error: this is a cosmiconfig error

`;

exports[`unknown plugin when there is no config to load, then no config is passed: when there is no config to load, then no config is passed 1`] = `

import configured from './configurable.macro'

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

// eslint-disable-next-line babel/no-unused-expressions
configured\`stuff\`


`;

exports[`unknown plugin works with function calls: works with function calls 1`] = `

import myEval from './fixtures/eval.macro'
const x = myEval('34 + 45')

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79


`;

exports[`unknown plugin works with import: works with import 1`] = `

import myEval from './fixtures/eval.macro'
const x = myEval\`34 + 45\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79


`;

exports[`unknown plugin works with require destructuring and aliasing: works with require destructuring and aliasing 1`] = `

const {css: CSS, styled: STYLED} = require('./fixtures/emotion.macro')
const red = CSS\`
  background-color: red;
\`

const Div = STYLED.div\`
  composes: \${red}
  color: blue;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const red = 'background-color: red;'
const Div = STYLED.div\`composes: background-color: red;
  color: blue;\`


`;

exports[`unknown plugin works with require destructuring: works with require destructuring 1`] = `

const {css, styled} = require('./fixtures/emotion.macro')
const red = css\`
  background-color: red;
\`

const Div = styled.div\`
  composes: \${red}
  color: blue;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const red = 'background-color: red;'
const Div = styled.div\`composes: background-color: red;
  color: blue;\`


`;

exports[`unknown plugin works with require: works with require 1`] = `

const evaler = require('./fixtures/eval.macro')
const x = evaler\`34 + 45\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79


`;
